File: Sinuka/src/Sinuka.Application/UseCases/Register/RegisterUseCase.cs
using System.Threading.Tasks;
using Sinuka.Core.Interfaces.Repositories;
using Sinuka.Core.Interfaces.Factories;
using Sinuka.Core.Interfaces.Services;
using Sinuka.Application.Interfaces;

namespace Sinuka.Application.UseCases.Register
{
    public class RegisterUseCase : IRegisterUseCase
    {
        private IRegisterPresenter _presenter;
        private readonly IUserRepository _userRepository;
        private readonly IUserFactory _userFactory;
        private readonly IEmailService _emailService;
        private readonly IUnitOfWork _unitOfWork;

        public RegisterUseCase(
            IUserRepository userRepository, 
            IUserFactory userFactory,
            IEmailService emailService,
            IUnitOfWork unitOfWork)
        {
            this._userRepository = userRepository;
            this._userFactory = userFactory;
            this._emailService = emailService;
            this._unitOfWork = unitOfWork;
        }

        public async Task Run(RegisterInput input)
        {
            if(await this._userRepository.FindUserByUsername(input.Username) is not null)
            {
                this._presenter.UsernameIsTaken();
                return;
            }

            if(await this._userRepository.FindUserByEmail(input.Email) is not null)
            {
                this._presenter.EmailIsTaken();
                return;
            }

            var user = this._userFactory.CreateUser(input.Username, input.Password, input.Email);

            await this._userRepository.AddUser(user);

            await this._unitOfWork.Save();

            await this._emailService.SendEmail(
                user.Email.Email, 
                $"localhost:8080/emailAddress/verifyEmail/{user.Email.VerificationString}", 
                "Verify Email"
            );

            this._presenter.UserCreated();
        }

        public void SetPresenter(IRegisterPresenter presenter) => this._presenter = presenter;
    }
}

File: Sinuka/Core/UseCases/RegisterUseCase.py
from ..Repositories.UserRepository import UserRepository
from ..Factories.UserFactory import UserFactory


class RegisterInput:
    username: str = None
    password: str = None
    email: str = None

    def __init__(self, username: str, password: str, email: str):
        self.username = username
        self.password = password
        self.email = email


class RegisterPresenter:
    def __init__(self):
        pass

    def existing_email_error(self):
        pass

    def existing_username_error(self):
        pass

    def user_created(self):
        pass


class RegisterUseCase:
    presenter: RegisterPresenter = None
    user_repo: UserRepository = None
    user_factory: UserFactory = None

    def __init__(self, user_repo: UserRepository, user_factory: UserFactory):
        self.user_repo = user_repo
        self.user_factory = user_factory

    def run(self, use_case_input: RegisterInput):
        if self.user_repo.find_user_by_username(use_case_input.username) is not None:
            self.presenter.existing_username_error()
            return

        if self.user_repo.find_user_by_email(use_case_input.email) is not None:
            self.presenter.existing_email_error()
            return

        user = self.user_factory.create_user(
            use_case_input.username,
            use_case_input.password,
            use_case_input.email
        )

        self.user_repo.add_user(user)

        self.presenter.user_created()

    def set_presenter(self, presenter: RegisterPresenter):
        self.presenter = presenter
